### PR TITLE
Add Scrapy v1.1.3 recipe.

### PR DESCRIPTION
**Pull request by rolando**

### CODE REVIEW

**Comment by rolando** [comment]
Appveyor builds for py27 succeed. Builds for py3* failed with:
```
[00:02:54] Installing scrapy.exe script to C:\Miniconda35-x64\envs\_build\Scripts
[00:02:54] writing list of installed files to 'record.txt'
[00:02:55] compiling .pyc files...
[00:03:08] Traceback (most recent call last):
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build_all\builder.py", line 200, in build
[00:03:08]     conda_build.api.build(meta.meta, config=config)
[00:03:08] AttributeError: module 'conda_build' has no attribute 'api'
[00:03:08] 
[00:03:08] During handling of the above exception, another exception occurred:
[00:03:08] 
[00:03:08] Traceback (most recent call last):
[00:03:08]   File "C:\Miniconda35-x64\Scripts\conda-build-all-script.py", line 11, in <module>
[00:03:08]     load_entry_point('conda-build-all==1.0.0', 'console_scripts', 'conda-build-all')()
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build_all\cli.py", line 90, in main
[00:03:08]     b.main()
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build_all\builder.py", line 263, in main
[00:03:08]     built_dist_location = self.build(meta, build_config)
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build_all\builder.py", line 204, in build
[00:03:08]     return bldpkg_path(build.build(meta.meta))
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build_all\build.py", line 34, in build
[00:03:08]     build_module.build(meta, post=None, need_source_download=True, **kwd)
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build\build.py", line 650, in build
[00:03:08]     preserve_egg_dir=bool(m.get_value('build/preserve_egg_dir')))
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build\post.py", line 194, in post_process
[00:03:08]     compile_missing_pyc(files)
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build\post.py", line 190, in compile_missing_pyc
[00:03:08]     coerce_pycache_to_old_style(compile_files, cwd=cwd)
[00:03:08]   File "C:\Miniconda35-x64\lib\site-packages\conda_build\post.py", line 166, in coerce_pycache_to_old_style
[00:03:08]     os.rename(fname, f + 'c')
[00:03:08] FileExistsError: [WinError 183] Cannot create a file when that file already exists: 'C:\\Miniconda35-x64\\envs\\_build\\Lib\\site-packages\\scrapy\\templates\\project\\module\\__pycache__\\__init__.cpython-35.pyc' -> 'C:\\Miniconda35-x64\\envs\\_build\\Lib\\site-packages\\scrapy\\templates\\project\\module\\__init__.pyc'
[00:03:08] Command exited with code 1
```

File: recipes/scrapy/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,100 @@
+{% set name = "Scrapy" %}
+{% set version = "1.1.3" %}
+{% set hash_type = "md5" %}
```
Comment (synapticarbors) [inline recipes/scrapy/meta.yaml]: I believe that sha256 is preferred: https://github.com/conda-forge/staged-recipes/wiki/Frequently-asked-questions#2-how-do-i-populate-the-hash-field

File: recipes/scrapy/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+build:
+  script: python setup.py install --single-version-externally-managed --record=record.txt
+  entry_points:
+    - scrapy = scrapy.cmdline:execute
+  number: 0
+
+# The requirements below shall match the requirements from the target package
+# version (check setup.py for changes).
+requirements:
+  build:
```
Comment (synapticarbors) [inline recipes/scrapy/meta.yaml]: I'm not sure that all of these are build time requirements. I think you might be able to just use `python` and `setuptools`.

File: recipes/scrapy/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    - python x.x
```
Comment (synapticarbors) [inline recipes/scrapy/meta.yaml]: I don't think you need to specify `x.x` for `python`. This is generally only used to ensure numpy ABI compatibility

File: recipes/scrapy/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    - scrapy.utils
+    - scrapy.xlib
+    - scrapy.xlib.tx
+
+  commands:
+    # Ideally, we will run tests here but we require to provide the tests
+    # requirements first.
+    - scrapy version
+
+about:
```
Comment (synapticarbors) [inline recipes/scrapy/meta.yaml]: Please add:

- `license_file: LICENSE`
- `dev_url: https://github.com/scrapy/scrapy`
- `doc_url: https://scrapy.org/doc/`

and possibly a more detailed description. See for example the recipe that Continuum uses in the package that's available through the default channel:

https://github.com/ContinuumIO/anaconda-recipes/blob/master/scrapy/meta.yaml#L38-L47

File: recipes/scrapy/bld.bat [hunk 5]
```diff
@@ -0,0 +1,2 @@
+"%PYTHON%" setup.py install --single-version-externally-managed --record=record.txt
```
Comment (synapticarbors) [inline recipes/scrapy/bld.bat]: I don't think the `bld.bat` file is required if you are using the `build: script:` section of the `meta.yaml`. That said, the recipe from Continuum uses separate `build.sh` and `bld.bat` files to delete some cryptography thing that I'm not 100% sure about:

https://github.com/ContinuumIO/anaconda-recipes/blob/master/scrapy/build.sh
https://github.com/ContinuumIO/anaconda-recipes/blob/master/scrapy/bld.bat

**Comment by synapticarbors** [comment]
@rolando I think the appveyor issue might be related to having the `bld.bat` in addition to the `build: script` section in the `meta.yaml` that I mentioned.

File: recipes/scrapy/meta.yaml (hunk 2 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: Right, fixed.

File: recipes/scrapy/meta.yaml (hunk 3 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: That leaked from a past recipe, fixed.

File: recipes/scrapy/meta.yaml (hunk 4 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: Done.

File: recipes/scrapy/bld.bat (hunk 5 above)
Comment (rolando) [inline recipes/scrapy/bld.bat]: Removed `bld.bat`.

File: recipes/scrapy/meta.yaml (hunk 1 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: Fixed.

**Comment by rolando** [comment]
@synapticarbors Thank you for your prompt feedback!

I made the suggested changes and build for py3[45] windows still fail. For some reason conda packages are downgraded:
```
             platform : win-32
        conda version : 4.1.11
    conda-env version : 2.5.2
  conda-build version : 1.21.14+0.g4dfebe9.dirty
       python version : 3.4.3.final.0
```

:thinking: I think I know why it's failing, the templates directory is included with .pyc files.

**Comment by synapticarbors** [comment]
@rolando No problem. Scrapy is one of my favorite projects and I use it in production for the majority of the web scraping I do. Thanks for making it available to the community.

 I'm not sure about the Windows build failures. Maybe one of the conda-forge members who is a bit more familiar with the build process can help if you can't resolve it.

**Comment by rolando** [comment]
That operation was removed from conda-build 20 days ago: https://github.com/conda/conda-build/commit/b9861cab8b2616768959d3c5518980146b63c213

It seems that we could try to remove `pyc`'s files in a `bld.bat` but I've decided to skip the build for win+py3k given the framework is not yet 100% functional in that stack.

File: recipes/scrapy/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  home: http://scrapy.org/
+  license: BSD
```
Comment (jakirkham) [inline recipes/scrapy/meta.yaml]: Could we please add the number of clauses (e.g. `BSD 3-Clauses` if it is 3-Clauses)?

File: recipes/scrapy/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    Scrapy is an open source and collaborative framework for extracting the
+    data you need from websites in a fast, simple, yet extensible way.
+  doc_url: http://scrapy.org/doc/
+  dev_url: https://github.com/scrapy/scrapy
+
+extra:
+  recipe-maintainers:
+    - rolando
+    - redapple
+    - kmike
```
Comment (jakirkham) [inline recipes/scrapy/meta.yaml]: Are you ok with being maintainers on this, @redapple @kmike?

File: recipes/scrapy/meta.yaml (hunk 6 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: Done!

File: recipes/scrapy/meta.yaml (hunk 7 above)
Comment (redapple) [inline recipes/scrapy/meta.yaml]: @jakirkham , I'm ok with this.
What are the duties in practice? does it give special permissions also?

**Comment by patricksnape** [comment]
Hmm, it would be really good to get Windows + Python 3.5 given that scrapy is a really popular package. That PYC error may have something to do with the entry point change in behaviour between Python 2 and 3? But it's confusing that it only fails on Python 3.5. Maybe something to do with the new zipapp support?

**Comment by synapticarbors** [comment]
@patricksnape According to the docs, Python 3 is not supported on windows:
https://doc.scrapy.org/en/latest/faq.html#what-python-versions-does-scrapy-support

File: recipes/scrapy/meta.yaml (hunk 7 above)
Comment (synapticarbors) [inline recipes/scrapy/meta.yaml]: @redapple Here is a list of the duties of a maintainer:

https://github.com/conda-forge/staged-recipes/wiki/Frequently-asked-questions#10-what-does-being-a-conda-forge-feedstock-maintainer-entail

You get full write privileges to the final feedstock.

**Comment by rolando** [comment]
@patricksnape I tracked error down to a pyc removal/rename attempted by conda build command. I'm not sure, but I could try running `conda update conda-build` in the appveyor build.

However, for now there is no a strong reason to fix the build for py3+win if the package will not work anyway.

**Comment by patricksnape** [comment]
@synapticarbors @rolando Ah sorry! I misunderstood - skip away then!

File: recipes/scrapy/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+  run:
+    - python
+    - twisted >=10.0.0
+    - w3lib >=1.15.0
+    - queuelib
+    - lxml
+    - pyopenssl
+    - cssselect >=0.9
+    - six >=1.5.2
```
Comment (patricksnape) [inline recipes/scrapy/meta.yaml]: Do these need to be pinned to patch specific versions?

Comment (rolando) [inline recipes/scrapy/meta.yaml]: That comes straight from scrapy's `setup.py`: https://github.com/scrapy/scrapy/blob/master/setup.py#L43

File: recipes/scrapy/meta.yaml (hunk 7 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: Removed kmike from the list as he is busy for the moment.

File: recipes/scrapy/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    - scrapy.settings
+    - scrapy.spidermiddlewares
+    - scrapy.spiders
+    - scrapy.utils
+    - scrapy.xlib
+    - scrapy.xlib.tx
+
+  commands:
+    # Ideally, we will run tests here but we require to provide the tests
+    # requirements first.
```
Comment (jakirkham) [inline recipes/scrapy/meta.yaml]: I'm a little confused by this comment. What is this suppose to mean? Do we need certain requirements to run the tests? If so, it is easy to add them. Or is there something else required here?

File: recipes/scrapy/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,96 @@
+{% set name = "Scrapy" %}
+{% set version = "1.1.3" %}
+{% set hash_type = "sha256" %}
+{% set hash_val = "fab9c5074bc484add855d2d072ff26fdc0ba5344858cc68178f4f4f39a6f2cce" %}
+package:
```
Comment (jakirkham) [inline recipes/scrapy/meta.yaml]: Might space out `package` from Jinja to make this a little more readable.

**Comment by synapticarbors** [comment]
I just noticed that Scrapy 1.2 was just released:

https://pypi.python.org/pypi/Scrapy/1.2.0

We should probably continue with this recipe as is and then update the feedstock once it is generated.

File: recipes/scrapy/meta.yaml (hunk 10 above)
Comment (rolando) [inline recipes/scrapy/meta.yaml]: Thanks for your suggestion, done!

**Comment by rolando** [comment]
@synapticarbors I had the same thought.

File: recipes/scrapy/meta.yaml (hunk 8 above)
Comment (patricksnape) [inline recipes/scrapy/meta.yaml]: It just that it seems a bit specific with things like the six pinning. But they are all `>=` so it's not a big concern.

**Comment by rolando** [comment]
Looks like the travis error is due to an internal error. The raw log states the build was OK:

```
===== scrapy-1.1.3-py35_0 OK =====

+/bin/bash -x -e /Users/travis/miniconda/conda-bld/test-tmp_dir/run_test.sh

+scrapy version

Scrapy 1.1.3

TEST END: scrapy-1.1.3-py35_0



travis_time:end:0c2074bd:start=1475689369521033000,finish=1475690751000000000,duration=1381478967000
[0K

[32;1mThe command "if [ -n "$GH_TOKEN" ]; then echo "Creating feedstocks from the recipe(s)."; git config --global user.name "Travis-CI on github.com/conda-forge/staged-recipes"; git config --global user.email "conda-forge@googlegroups.com"; source ./.CI/create_feedstocks; else echo "Building all recipes."; source ./.CI/build_all; fi" exited with 0.[0m



Done. Your build exited with 0.

/Users/travis/build.sh: line 109: shell_session_update: command not found
```

**Comment by jakirkham** [comment]
That's because it is not using this right image. 😒 Will try closing and reopening to see if that fixes it.

xref: https://travis-ci.org/conda-forge/staged-recipes/builds/165297322#L4

**Comment by rolando** [comment]
Travis failed with

```
Finding recipes merged in master and removing them from the build.
Removing recipe: example
./.CI/build_all: line 28: conda-build-all: command not found
```

**Comment by jakirkham** [comment]
Yeah that's obscuring the real error, which is probably we got a 502 error when trying to install `conda-build-all`. In any event, went away with a restart.

**Comment by jakirkham** [comment]
Thanks everyone. Let's get out there. 😄

**Comment by rolando** [comment]
@jakirkham @synapticarbors @patricksnape @redapple thank you all!

File: recipes/scrapy/meta.yaml (hunk 9 above)
Comment (redapple) [inline recipes/scrapy/meta.yaml]: @rolando , i'm not sure I'm getting this.
If I search for `mitmproxy` or `pytest-twisted` on https://anaconda.org/search, I do find them, though mostly source distribution from the PyPI mirror channel.
Is that not enough? (sorry for the newbie question, I'm not a conda user myself)

**Comment by redapple** [comment]
Thanks @rolando !

Commenting on the above questions, the main reason Scrapy is not officially supported on Windows with Python 3 is that we cannot test it (because of Twisted not officially supported on Win+Py3)

And a really nice thing about anaconda is that Twisted for example [is available on anaconda](https://anaconda.org/anaconda/twisted/files?version=16.4.1) with Py34 and Py35 on Win32 and Win64.

If we can do it, I would like to have Scrapy+Python3+Windows available, yet only on anaconda, hence also be able to run the tests.

**Comment by rolando** [comment]
For the record, I tracked down the win build error to the Scrapy's package manifest not ignoring compiled files, so that caused to include `.pyc` files inside the templates directory which were not cleaned up properly and caused conda to fail when compiling `.pyc` files. See this PR for recipe changes https://github.com/conda-forge/scrapy-feedstock/pull/5/files

Alternatively, newer versions of conda provides a ``skip_compile_pyc`` flag which may help to address this issue in other packages. See http://conda.pydata.org/docs/building/meta-yaml.html#skip-compiling-some-py-files-into-pyc-files

**Pull request merged**